server/text_generation_server/models/causal_lm.py

```diff
@@ -265,52 +283,59 @@ def to_tensors(ind): return (torch.tensor(ind[0], device=device), torch.tensor(i
             del b.input_ids
-        src = pad_tensors(src, paddings, seq_dim, pad_token_id)
+        src = pad_tensors(src, extra_padding, seq_dim, pad_token_id)
         src = shift_all(src, seq_dim, offsets)
         input_ids = prepare_memory(new_bs, src[target_batch_idx], inplace)
         input_ids = move_data(input_ids, 1, indices, src)
 
         src = [b.attention_mask for b in batches]
         for b in batches:
             del b.attention_mask
-        src = pad_tensors(src, paddings, seq_dim, 0)
+        src = pad_tensors(src, extra_padding, seq_dim, 0)
         src = shift_all(src, seq_dim, offsets)
         attention_mask = prepare_memory(new_bs, src[target_batch_idx], inplace)
         attention_mask = move_data(attention_mask, 1, indices, src)
 
         src = [b.position_ids for b in batches]
         for b in batches:
             del b.position_ids
-        src = shift_all(src, seq_dim, offsets)
         position_ids = prepare_memory(new_bs, src[target_batch_idx], inplace)
         position_ids = move_data(position_ids, 1, indices, src)
 
-        past_key_values = []
-        for layer_num in range(num_layers):
-            src = [b.past_key_values[layer_num][0] for b in batches]
-            src = pad_tensors(src, paddings, key_dim, 0)
-            src = shift_all(src, key_dim, offsets)
-            updated_key = prepare_memory(new_bs * chunk_size, src[target_batch_idx], inplace)
-            updated_key = move_data(updated_key, chunk_size, indices, src)
+        src = None
+        src_keys = [[b.past_key_values[layer_num][0] for layer_num in range(num_layers)] for b in batches]
+        src_values = [[b.past_key_values[layer_num][1] for layer_num in range(num_layers)] for b in batches]
+        for b in batches:
+            del b.past_key_values
 
-            src = [b.past_key_values[layer_num][1] for b in batches]
-            src = pad_tensors(src, paddings, value_dim, 0)
-            src = shift_all(src, value_dim, offsets)
-            updated_value = prepare_memory(new_bs * chunk_size, src[target_batch_idx], inplace)
-            updated_value = move_data(updated_value, chunk_size, indices, src)
+        src_keys = [torch.stack(src) for src in src_keys]
+        htorch.core.mark_step()
+        src_keys = pad_tensors(src_keys, extra_padding, key_dim, 0)
+        htorch.core.mark_step()
+        src_keys = shift_all(src_keys, key_dim, offsets)
+        htorch.core.mark_step()
+        src_keys = [[t.squeeze(0).clone() for t in torch.split(src, 1)] for src in src_keys]
+        htorch.core.mark_step()
 
-            past_key_values.append((updated_key, updated_value))
-            for b in batches:
-                b.past_key_values[layer_num] = None
+        dst_keys = [prepare_memory(new_bs * chunk_size, prev, inplace) for prev in src_keys[target_batch_idx]]
+        dst_keys = [move_data(dst_keys[layer_num], chunk_size, indices, [src[layer_num] for src in src_keys]) for layer_num in range(num_layers)]
```

Comment:
is it more performant to loop over src_keys and src_values multiple times and place mark steps? instead to what we had (single loop over layers)

Reply:
Unfortunately those mark_steps are needed to limit the number of recipes.
In previous approach we called shift 80*2 times and op accumulation + graph starting overhead was dominating run time.

Reply:
I'll double check... Both pad and shift are calling mark_step on their own so in theory we could get rid of some of them...